Test/Unit/Controller/Cart/EmailTest.php

```diff
@@ -0,0 +1,182 @@
+     */
+    public function testNoEmail()
+    {
+        $this->checkoutSession->method('getQuote')->willReturn($this->quote);
+        $this->request->method('getParam')->willReturn('');
+        $exception = new LocalizedException(__('No email received.'));
+        $this->bugsnag->expects($this->once())
+                    ->method('notifyException')
+                    ->with($this->equalTo($exception));
+        $this->currentMock->execute();
+    }
+
+    /**
+     * This doesn't really seem to test invalid emails as the validateEmail method
+     * needs to be mocked. Generally just making sure that if validateEmail comes
+     * back false we error handle properly
+     * @covers Email::execute
+     */
+
+    public function testInvalidEmail()
+    {
+        $this->checkoutSession->method('getQuote')->willReturn($this->quote);
+        $this->request->method('getParam')->willReturn('invalidemail');
+        $this->cartHelper->method('validateEmail')->willReturn(false);
+        $exception = new LocalizedException(__('Invalid email: %1', 'invalidemail'));
+        $this->bugsnag->expects($this->once())
+                    ->method('notifyException')
+                    ->with($this->equalTo($exception));
+        $this->currentMock->execute();
+    }
+
+    public function testOtherException()
+    {
+        $this->bugsnag->expects($this->once())->method('notifyException');
+        $this->currentMock->execute();
+    }
+
+    /**
+     * @test
+     * @covers Email::execute
+     */
+//
+    public function testExecuteHappyPath()
+    {
+        $this->checkoutSession->method('getQuote')->willReturn($this->quote);
+        $this->cartHelper->method('validateEmail')->willReturn(true);
+        $this->request->method('getParam')->willReturn('example@email.com');
+        $this->checkoutSession->expects($this->once())->method('getQuote');
+        $this->cartHelper->expects($this->once())->method('quoteResourceSave');
+        $this->currentMock->execute();
+    }
+
+    private function initRequiredMocks()
+    {
+        //objects needed for constructor
+        $this->context = $this->createMock(Context::class);
+        $this->checkoutSession = $this->createMock(CheckoutSession::class);
+        $this->customerSession = $this->createMock(CustomerSession::class);
+
+        $this->bugsnag = $this->createMock(Bugsnag::class);
+
+        $this->cartHelper = $this->createMock(CartHelper::class);
+
+        //objects needed for method
+        $this->quote = $this->createMock(Quote::class);
+        $this->quote->method('getId')->willReturn('1234');
+
+        $this->request = $this->createMock(RequestInterface::class);
+
+        //methods
+//        $this->checkoutSession->method('getQuote')->willReturn($this->quote);
+        $this->cartHelper->method('quoteResourceSave')->willReturn(null);
+        $this->context->method('getRequest')->willReturn($this->request);
+    }
+
+    private function initCurrentMock()
```

Comment:
Nit: let's move this to the top. When reading a test file all initializations etc tend to be at the top usually which makes it easy to understand what thebtest Is doing. Let's stick with that pattern.